NEXUS-4520 et al: undefined variables

cstamas opened the pull request:
This pull request, same as issue, is a bit "mish-mash".

Original intent of issue was to fix the lost capability to honor env vars and system properties to get into appcontext that has prefix "PLEXUS_" and "plexus." accordingly.

As the issue itself notes, this was already fixed and is in core.

This pull request contains:
- fix from nexus-mock module that was during build producing a folder "${nexus-work}" next to pom.xml
- fir for IT that was creating a folder "storage" next to pom.xml
- cleaning up git ignores
- IT for verifying the NEXUS-4520 _is fixed_ for case with system properties

cstamas committed:
NEXUS-4520: Fix for "${nexus-work}" appear if nexus-mock folder after running tests.

This is not in direct relation to NEXUS-4520 actually, but was fixed.
What was happening, that Mock module boots nexus in a "in situ" way (as opposed to
ITs where "cleanroom" boot happens in isolated classloader). This is needed to be able
to "mock" plexus resources properly. But this also means, that classloader executing
the test is having a "priority" over actual nexus' webapp classloader (jetty term).
But, due to a ahem-load of dependencies, almost all of Nexus is "lifted" into Test
UT. Hence, plexus-ehcahce, used by Security is there also, and it uses classloader
of it's own classes to locate the ehcache config on classpath. This makes
WEB-INF/classes invisible to it (it's in child classloader, not in test classloader),
hence creates a tmp folder for ehcahce using wrong configuration.

Fix for now is simply "lifting" the proper ehcache.xml config file into test
resources, but ultimately, the plexus-ehcache needs an update.

cstamas committed:
NEXUS-4520: updating the "fallback" default ehcache to latest one.

cstamas committed:
NEXUS-4520: Fixed nasty IT that was puking into module folder

cstamas committed:
NEXUS-4520: fixing migration plugin wrt some not updated properties

cstamas committed:
Removing unneeded ignores

cstamas committed:
NEXUS-4520: IT verifying that "plexus." prefixed system properties are picked up. Same
code is used for "PLEXUS_" prefixed env vars, but this IT does not check that for now.

nabcos commented on nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4520/Nexus4520PlexusPrefixedVariablesArePickedUpIT.java:
```diff
@@ -0,0 +1,82 @@
+ * TODO
+ *
+ * @author: cstamas
+ */
```
javadoc missing

nabcos commented:
aside from javadoc TODO +1

cstamas committed:
NEXUS-4520: added license header and doc'd what is happening here

cstamas commented:
Javadoc + license header of the new IT fixed.

peterlynch commented on nexus/nexus-mock/src/test/resources/ehcache.xml:
```diff
@@ -0,0 +1,372 @@
+            />
+
+    <!--
+    DiskStore configuration
```
Do people like to have this massive boiler plate comment in this file? - makes hard to read - why not remove it and add a link to the online docs instead?

cstamas replied:
Actually, I just realized that this file here (in nexus-mock) is unneeded. The problem was that ehcache-default.xml in nexus-proxy was outdated. Meaning, this file should be removed.

And wrt your "boiler plate" comment: these files are _delivered within jar_, and are just "fallbacks" when no real config is found. They are not meant for end users editing (and they are anyway unable to edit these since they are packed into nexus-proxy JAR).

But, these files do originate from vanilla ehcache distro, meaning, leaving the as is and having only applied "our" (nexus specific) changes is handy, since you can track the changes using diff.

peterlynch replied:
ok sounds good.

peterlynch commented on nexus/nexus-proxy/src/main/resources/ehcache-default.xml:
```diff
@@ -32,12 +34,31 @@
+    <cache name="shiro-activeSessionCache"
+            maxElementsInMemory="10000"
+            eternal="true"
+            overflowToDisk="false"
```
again would like to remove the massive boiler plate comment in this file

peterlynch commented on nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4520/Nexus4520PlexusPrefixedVariablesArePickedUpIT.java:
```diff
@@ -0,0 +1,102 @@
+/**
+ * IT for issue NEXUS-4520, covering the "System properties" case. In short, before Nexus is booted, it sets a
+ * system propery like {@code plexus.foo=bar} (simulating option -Dplexus.foo=bar in wrapper.conf for example) and
+ * then checks for the presence of {@code foo=bar} key-value pair in AppContext.
```
thx for the comment

peterlynch commented:
+1

cstamas committed:
NEXUS-4520: After chatting with Peter, I realized that this added file
-- while did fix the creation for "${nexus-work}" folder is actually
just a symptom treatment, since the initial cause was outdated "fallback"
ehcache config present in nexus-proxy (packed into jar).
That file is already fixed.

cstamas committed:
NEXUS-4520: applied Matcher change as suggested, but also reduced the matched string to
avoid future IT failures that will fail if AppContext changes it's "dump" format. Now the
match is done only against the presence of the key-value pair.

cstamas commented:
Matcher changed to `containsString`, as suggested, but also reduced the matched string, to not have IT failures if AppContext "dump format" (something totally unrelated to Nexus) changes in future.
